### PR TITLE
add ZIP distribution

### PR DESCRIPTION
**Pull request by alicanb**
questions for reviewers:
- This definitely needs more tests, looking for suggestions
- I choose subclassing `TorchDistribution` directly. Other option was creating `ZeroInflatedDistribution` abstract class, for easily implementing ZINB as well, but it wasn't generic enough to provide enough code reuse imo. Do you agree?

### CODE REVIEW

**Comment by fritzo** [comment]
cc @OptimusLime

File: pyro/distributions/zip.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all, lazy_property
+
+from pyro.distributions import TorchDistribution
+
+
+class ZIP(TorchDistribution):
```
Comment (fritzo) [inline pyro/distributions/zip.py]: nit: I'm tempted to name this `ZeroInflatedPoisson` and let users `import ZeroInflatedPoisson as ZIP`

File: pyro/distributions/zip.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        batch_shape = self.gate.shape
+        event_shape = torch.Size()
+        super(ZIP, self).__init__(batch_shape, event_shape, validate_args)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        gate, rate, value = broadcast_all(self.gate, self.rate, value)
+        log_prob = (-gate).log1p() + (rate.log() * value) - rate - (value + 1).lgamma()
+        log_prob[value == 0] = (gate[value == 0] + (1 - gate[value == 0]) * torch.exp(-rate[value == 0])).log()
```
Comment (fritzo) [inline pyro/distributions/zip.py]: Let's pull out the common subexpression
```
zeros = (value == 0)
```

File: pyro/distributions/zip.py (hunk 1 above)
Comment (jpchen) [reply pyro/distributions/zip.py]: +1 for more descriptive names, it's also consistent with the rest of the distribution names

**Comment by alicanb** [comment]
Yep, good to go!

**Pull request merged**